pkg/controller/seed-controller-manager/etcdbackup/etcd_backup_controller.go

```diff
@@ -0,0 +1,348 @@
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+const (
+	ControllerName     = "kubermatic_etcd_backup_controller"
+	defaultClusterSize = 3
+
+	// DeleteAllBackupsFinalizer indicates that the backups still need to be deleted in the backend
+	DeleteAllBackupsFinalizer = "kubermatic.io/delete-all-backups"
+)
+
+type BackendOperations interface {
+	takeSnapshot(ctx context.Context, log *zap.SugaredLogger, fileName string, cluster *kubermaticv1.Cluster) error
+	uploadSnapshot(ctx context.Context, log *zap.SugaredLogger, fileName string) error
+	deleteSnapshot(ctx context.Context, log *zap.SugaredLogger, fileName string) error
+	deleteUploadedSnapshot(ctx context.Context, log *zap.SugaredLogger, fileName string) error
+}
+
+// Reconciler stores necessary components that are required to create etcd backups
+type Reconciler struct {
+	log        *zap.SugaredLogger
+	workerName string
+	ctrlruntimeclient.Client
+	BackendOperations
+	clock    clock.Clock
+	recorder record.EventRecorder
+}
+
+// Add creates a new Backup controller that is responsible for
+// managing cluster etcd backups
+func Add(
+	mgr manager.Manager,
+	log *zap.SugaredLogger,
+	numWorkers int,
+	workerName string,
+	snapshotDir string,
+	s3Endpoint string,
+	s3BucketName string,
+	s3AccessKeyID string,
+	s3SecretAccessKey string,
+) error {
+	log = log.Named(ControllerName)
+	client := mgr.GetClient()
+
+	reconciler := &Reconciler{
+		log:        log,
+		Client:     client,
+		workerName: workerName,
+		recorder:   mgr.GetEventRecorderFor(ControllerName),
+		BackendOperations: &s3BackendOperations{
+			snapshotDir:       snapshotDir,
+			s3Endpoint:        s3Endpoint,
+			s3BucketName:      s3BucketName,
+			s3AccessKeyID:     s3AccessKeyID,
+			s3SecretAccessKey: s3SecretAccessKey,
+		},
+		clock: &clock.RealClock{},
+	}
+
+	ctrlOptions := controller.Options{
+		Reconciler:              reconciler,
+		MaxConcurrentReconciles: numWorkers,
+	}
+	c, err := controller.New(ControllerName, mgr, ctrlOptions)
+	if err != nil {
+		return err
+	}
+
+	return c.Watch(&source.Kind{Type: &kubermaticv1.EtcdBackup{}}, &handler.EnqueueRequestForObject{})
+}
+
+func (r *Reconciler) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	log := r.log.With("request", request)
+	log.Debug("Processing")
+
+	backup := &kubermaticv1.EtcdBackup{}
+	if err := r.Get(ctx, request.NamespacedName, backup); err != nil {
+		if kerrors.IsNotFound(err) {
+			return reconcile.Result{}, nil
+		}
+		return reconcile.Result{}, err
+	}
+
+	cluster := &kubermaticv1.Cluster{}
+	if err := r.Get(ctx, types.NamespacedName{Name: backup.Spec.Cluster.Name}, cluster); err != nil {
+		return reconcile.Result{}, err
+	}
+
+	log = r.log.With("cluster", cluster.Name, "backup", backup.Name)
+
+	// Add a wrapping here so we can emit an event on error
+	result, err := kubermaticv1helper.ClusterReconcileWrapper(
```

Comment:
We can probably use this to set a condition with the last backup status for the cluster.

Reply:
Not with `ClusterReconcileWrapper`, I don't think. It sets the condition when the reconcile method returned nil,nil -- which the etcd backup controller's reconcile method never will for backup configs with a schedule until the backup config is deleted (because with a schedule set, I'm always returning a "requeueAfter: [time to next scheduled backup]" result, not nil). That's why I thought that aspect of `ClusterReconcileWrapper()` doesn't make too much sense here -- it seems to be basically meant for controllers that reconcile something once per cluster and are then done with it.

If we really wanted to set a condition explicitly every time a backup succeeds, I'd have to add explicit support for it.